File: Unit-4-Object-Oriented-Programming/Day-3-Playing-Card-Example-V2/Day-3-Playing-Card-Example-V2/PlayingCard.cs
﻿
using System;
using System.Collections.Generic;

namespace Day_3_Playing_Card_Example_V2
{
    // This is class to represent a simple PlayingCard
    // It's in the same namespace as the application (for now and for simplicity)
    //
    // access - public - any one can access and use
    //          private - only members of this class can access and use
    //
    // A class contains data and methods to operate and manipulate the data

//  access type  name
    public class PlayingCard
    {
        /*********************************************************************
         * Data members (attributes, properties, variables)
         *********************************************************************/

        // Define instance variables for the class (unique attributes for each object)
        //
        // Instance variables are private to implement Encapsulation
        //      (protect data from access outside the class)

        private int    cardValue;
        private string cardSuit;
        private string cardColor;

        // To provide access so other can "see" or data members
        // We can provide "properties" to do that

        // A property is a public variable to allow access to private data
        //   using "getters" and "setters"
        //
        // A getter will return the value in a data member
        // A setter will allow the changing of data member
        //
        // A property is typically name the same as the variable in PascalCase

        // Define properties for access to data members
        public int CardValue  // name is the data member in PascalCase
        {
            get { return cardValue; }   // getter - return the value in cardValue
            set                         // setter - set cardValue to value used when assigning
            {
                cardValue = value;      //          value is keyword representing the value assigned
                ValidateValue();
            }   
                                               
        }
        public string CardSuit  // name is the data member in PascalCase
        {
            get { return CardSuit; }    // getter - return the value in cardSuit
            set                         // setter - set cardSuit to value used when assigning
            {
                cardSuit = value;       // value is keyword representing the value assigned 
                setColor();             // set color bed on suit;
            }  
                   
        }
        public string CardColor  // name is the data member in PascalCase
        {
            get { return cardColor; }   // getter - return the value in cardValue

            // Note: No setter - we do not want the user to set the color
                  
        }
        /*********************************************************************
         * Method members (functions that operate on the class data)
         *********************************************************************/

        // Every class has at least one constructor
        //
        // Constructors are responsible for initializing data in the object
        //
        // Constructors:   public, same name as class, no return type, may have parameters

        // Constructor to initialize all instance variables
        // Note: Color is NOT provided by the user
        //       we decide color based on suit
        public PlayingCard(int theValue, string theSuit)
        {
            cardValue = theValue;  // initialize value to value passed 
            ValidateValue();       // Check to be sure value is valid

            cardSuit  = theSuit;   // initialize suit to suit passed  
            // Check to see if the suit is valid... if not, set to default suit - "Spades"
            if (!ValidateSuit())
            {
                cardSuit = "Spades";
            }
            // cardSuit must be set to a valid suit before we try to set the color
            setColor();            // set cardColor based on suit
        }

        // a Copy constructor to create a copy of a PlayingCard from a PlayingCard
        //   (aka "deep copy")
        // Avoids two references to same object when creating an object from an object
        //
        // A copy constructor receives and object of the class as a parameter
        // and initialize using the data members in that object

        public PlayingCard(PlayingCard sourceCard)
        {
            cardColor = sourceCard.cardColor;
            cardSuit  = sourceCard.cardSuit;
            cardValue = sourceCard.cardValue;
        }

        /**************************************************************************
         * Class methods define the behavior of the class
         *************************************************************************/
        // This method is private so that only members of the class can use it
        //  i.e. users of the class can't use it or even see
        private void setColor()
        {
            Dictionary<string, string> suitColors = new Dictionary<string, string>();

            suitColors.Add("Clubs"   , "Black");
            suitColors.Add("Hearts"  , "Red");
            suitColors.Add("Diamonds", "Red");
            suitColors.Add("Spades"  , "Black");

            cardColor = suitColors[cardSuit];  // Set color based on suit
        }  // End of SetColor)

        private bool ValidateSuit()
        {
            // Check to see if the suit provided by the user is valid or not
            switch (cardSuit)
            {
                // Stacked case are like an series of ==/or
                case "Spades":    // If the
                case "Clubs":     //    suit given
                case "Hearts":    //       is one
                case "Diamonds":  //          we want...
                {
                    return true;  // return true
                }
                default:          // if not one we want...
                {
                    return false; // return false
                }
            }
        }  // End of ValidateSuit()

        // Check the cardValue we were given and set to default value if invalid
        public void ValidateValue()
        {
            // If the card value is not between a Joker (0) and a King (13)
            if (cardValue < 0 || cardValue > 13)
            {
                cardValue = 0;  // Set the value to a Joker (default value)
            }
        }

        /**************************************************************************************
         * Method overrides to have class behave way we want not the default way
         *
         * System methods you should override:
         *
         *     public string ToString()         - Return a string representation of an object
         *     public bool   Equals(Object obj) - Compares the content of objects
         *                                        The default .Equals() compares locations
         *                                            NOT the content
         *     public int    GetHashCode()      - Generates a unique for the object using the
         *                                        data members in the object.
         *                                        A HashCode is used in several places in C# to
         *                                        to determine if two objects are equal.
         *                                        The default GetHashCode() method generates the
         *                                        value using the location of the object not the content
         *                                        the values should be equal if the content is equal,
         *                                            but won't be
         **************************************************************************************/

        // Override the default ToString() method
        // We MUST be sure the method signature matches the method we are overriding
        // An override is substituting your processing for the default processing

        // Since the default ToString() returns a string and receives no parameters
        // The override must do the same

        // the override keyword, tells C# this is an override for a default behavior
        //                       C# will check to be sure your override signature
        //                                           matches the default

        // A method signature identifies the method and contains:
        //     
        //   1. Return type
        //   2. a method name
        //   3. parameter

        // ToString() returns an object of the class as a string
        public override string ToString()  
        {
            return $"PlayingCard: Value={cardValue}, Color={cardColor}, Suit={cardSuit}";
        }

        // .Equals override to compare the contents of our objects
        // type override (and maybe space) it should show you the format for overrides
        // Note: .Equals receives a generic object as a parameter
        //       you must either cast the object to class at every use
        //        ot define an object of the cass to use in teh method
        public override bool Equals(object otherObject)
        {
            // Define a reference to the generic we are passed so we can use it
            PlayingCard theOtherCard = (PlayingCard)otherObject;

            // If all data members between the two objects are equal, the objects are equal
            // this. represents the object to left of the . when method was called:
            //     card1.Equals(card2) - .this represents card1, otherObject  represents card2
            if (theOtherCard.cardColor  == this.cardColor
            &&  theOtherCard.cardSuit   == this.cardSuit
            &&  theOtherCard.cardValue  == this.cardValue)
            {
                return true;
            }
            return false; ;
        } // End of .Equals()

        // Override the GetHashCOde() to generate a unique value for data in out class
        //
        // A HashCode can be easily created using the sum of:
        //
        //       numeric-values * prime-number
        //       string, bool, object - use the GetHashCode() defined for them
        //
        public override int GetHashCode()
        {
            return cardValue * 17 + cardColor.GetHashCode() + cardSuit.GetHashCode();
        }
    }  // End of PlayingCard class
} // End of namespace

File: Unit-5-Java/00-Playing-Card-From-CSharp-V1-2/src/PlayingCard/PlayingCard.java
// namespace Playing_Card_Poker_Game;  // C# groups things in namespaces
package PlayingCard; // Java groups things in package - package name must match the folder name for the package

import java.util.Objects;
// Just like in C#. Java class names are in PascalCase
// When you see a name starting in uppercase - it is a class
// When you see a name starting in lowercase - it is a variable or object
public class PlayingCard
{
    /***************************************************************************************************
     * Define constants to represent card attributes
     *
     * public is OK since they are constants and cannot be changed
     *
     * static so it can be referenced using the class name. i.e. no object required
     *
     * enum - define a set of constant values that may be referenced as a data type
     *        allows the assign of a word to a constant value to limit the values in a variable
     *        make it easier to code (get rid using "magic" numbers to represent data
     *        used as data-types - define variable as enums, parameters as enum - anywhere a variable is allowed
     *        Java will ensure that an enum type only has values valid for the enum (we don't have to check)
     *        enum are actual integer values starting at 0 inside C#
     ***************************************************************************************************/
    // Making the enum public allows application programs to use the enum as constant too

    public enum CardColor
    {  // define words to represent allowable card colors (instead of String)
        // 0  , 1  
        Black, Red                 // These are the only valid values C# will allow
    }  // C# required a ; at closing brace of enum; Java doesn't

    public enum CardSuit
    {   // public is OK since they are constants and cannot be changed
        Spade, Club, Heart, Diamond, Joker  // These are the only valid values C# will allow
    }

    public enum CardValue
    {  // Using the fact that enums are really integers inside value to name our values
        //  0    1    2     3      4    5     6     7      8      9    10   11    12     13    14
        Joker, One, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Jack, Queen, King, Ace
    }

    /***************************************************************************************************
     * Define constants to represent defaults for card attributes
     *
     * protected means members of this class and members of any subclass can access the data / method
     *           directly without having to use methods to do so
     *
     * protected so subclasses may access is OK since they are constants and cannot be changed
     *
     * static so it can be referenced using the class name. i.e. no object required
     *
     *   C#: readonly so it can't be changed - const is Ok too 
     * Java: final so it can't be changed
     ***************************************************************************************************/
//                            datatype  variable-name    = initial-value  ;
    protected static final CardValue DEFAULTCARDVALUE = CardValue.Joker;   // enum data-type for value
    protected static final CardColor DEFAULTCOLOR     = CardColor.Black;
    protected static final CardSuit  DEFAULTSUIT      = CardSuit.Joker;
    /***************************************************************************************************
     * Member data
     *
     * private to protect as prescribed by encapsulation - method must be used to access the data 
     * 
     *   C#: private variable name starts with _ by convention
     * Java: has no convention for naming private variables
     ***************************************************************************************************/

    private CardValue value;  // use enum for data type- C# will enforce allowable values
    private CardColor color;  // use enum for data type- C# will enforce allowable values
    private CardSuit  suit;   // use enum for data type- C# will enforce allowable values
    //   C#: bool
    // Java: boolean
    private boolean   faceUp; // true if card is face up; false if not   

    /***************************************************************************************************
     * Default constructor - in case it is needed
     ***************************************************************************************************/

    public PlayingCard()
    {
        value = DEFAULTCARDVALUE;
        suit = DEFAULTSUIT;
        setColor(suit);     // Set color in object based on suit - NOTE: suit must have a value before this
        faceUp = false;
    }
    /***************************************************************************************************
     * 2-arg (CardValue and CardSuit) constructor
     ***************************************************************************************************/
    public PlayingCard(CardValue value, CardSuit suit)
    {
        this.value = value;   // Set value in object to value passed as argument
        setColor(suit);        // Set color in object based on suit passed as argument
        this.suit = suit;     // Set suit in object to suit passed as argument
        faceUp = false;
    }
    /***************************************************************************************************
     * 2-arg (int and CardSuit) constructor
     ***************************************************************************************************/
    public PlayingCard(int value, CardSuit suit)
    {
        // to convert an int to enum
        this.value = setValue(value); // Set value in object based on int value passed as argument
        setColor(suit);               // Set color in object based on suit passed as argument
        this.suit = suit;             // Set suit in object to suit passed as argument
        faceUp = false;
    }
    /***************************************************************************************************
     * C# Only: Properties - Used for others to access private data via getters and setters
     * Java does not support properties
     * Java requires specific getter and setter methods for each private variables
     **************************************************************************************************/
    /***************************************************************************************************
     * Standard  getter and setter methods
     ***************************************************************************************************/
    public CardValue getValue() {
        return value;
    }

    public void setValue(CardValue value) {
        this.value = value;
    }

    public CardColor getColor() {
        return color;
    }

    public void setColor(CardColor color) {
        this.color = color;
    }

    public CardSuit getSuit() {
        return suit;
    }

    public void setSuit(CardSuit suit) {
        this.suit = suit;
    }

    public boolean isFaceUp() {
        return faceUp;
    }

    public void setFaceUp(boolean faceUp) {
        this.faceUp = faceUp;
    }

    /***************************************************************************************************
     *   C#: MethodNames are Pascal case (start with uppercase letter)
     *       Opening brace for the method is on the next line from method signature
     *       (same if true for any statement that hase a block - if, while, switch etc)
     *
     * Java: methodNames are in camelCase (start with lowercase letter)
     *       (except constructor which always has same nae as class (PascalCase)
     *
     *       Opening brace for the method is on the same line as method signature
     *       (same if true for any statement that hase a block - if, while, switch etc)
     ***************************************************************************************************/
    /***************************************************************************************************
     * additional getter and setter methods
     ***************************************************************************************************/

    // Return the integer value assigned to the card - value is an enum called CardValue
    public int getIntValue() {   // Return integer value of CardValue enum in a PlayingCard
     // return (int) value;  //   C#: cast the enum to an int to get its integer value
        return value.ordinal(); // Java: .ordinal() method return the enums int value
    }

    public CardValue setValue(int ivalue) {  // Set the CardValue based on an int value
        /***********************************************************************************************************
         * switch is alternative to a set of nested if-then-else-if statements
         * switch is followed by a series of case statements which are the value you want to check in the variable
         * when a case is true the statements following the case AND ALL OTHER CASES are run too
         *        unless you have break at end of the case
         *        when a case is true all statements in the switch following the case are run until
         *        it hits a break statement or the end of the switch - no other cases are checked
         *
         * stacking cases to simulate an equals-or condition
         * There is not a way to simulate an equals-and in a switch
         *
         * switch(variable) { - check the value in this variable
         *      case value:  - if the value for the switch equales the case value
         *           statements-to-run if value in the case matches the Ace in the swutch
         *           break; - exit switch - if missing you will fall into next case
         *                    not required if case issues a return statement at end of case
         *      default: - if none of the cases were true
         *           statement-to-run if none of cases were true
         *           break; - exit switch -   C#: required on default case even though it is last
         *                                  Java: does not require break on default case
         * }
         *************************************************************************************************************/
        switch (ivalue)  { // check ivlue
            case 1:                      // if ivalue == 1
                return CardValue.Ace;    //     do this
            case 2:                      // if ivalue == 2
                return CardValue.Two;    //      do this
            case 3:                      // ivalue == 3
                return CardValue.Three;  //      do this
            case 4:
                return CardValue.Four;
            case 5:
                return CardValue.Five;
            case 6:
                return CardValue.Six;
            case 7:
                return CardValue.Seven;
            case 8:
                return CardValue.Eight;
            case 9:
                return CardValue.Nine;
            case 10:
                return CardValue.Ten;
            case 11:
                return CardValue.Jack;
            case 12:
                return CardValue.Queen;
            case 13:
                return CardValue.King;
            default:
                return CardValue.Joker;
        }
    }

    private void setColor(CardSuit suit) {  // Set the color based on the suit of the object
        // if (suit == cardSuit.Spade or suit == cardSuit.Club) - stack cases does this
        switch (suit) {
            case CardSuit.Spade:       // Stacking cases simulates an equal or condition
            case CardSuit.Club:
                this.color = CardColor.Black;
                break;      // break is required so we don't fall through to the next case - we exit the switch
            case CardSuit.Diamond:
            case CardSuit.Heart:
                this.color = CardColor.Red;
                break;      // break is required so we don't fall through to the next case - we exit the switch
            default:
                this.color = DEFAULTCOLOR;  // break is optional here in Java
                break;
        }
    }

    public void flip()  { // reverse if card is face up or down
        faceUp = !faceUp;
    }

    /***************************************************************************************************
     * Object super class overrides
     *
     *      toString()    - represent class data as a String
     *      equals()      - determine if contents of two object of the class are equal
     *      hashCode()    - generate hash code based on unchanging data members
     * 
     ***************************************************************************************************/
    @Override
    public String toString() {
        return "PlayingCard{" +
                "value=" + value +
                ", color=" + color +
                ", suit=" + suit +
                ", faceUp=" + faceUp +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof PlayingCard that)) return false;
        return faceUp == that.faceUp && value == that.value && color == that.color && suit == that.suit;
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, color, suit, faceUp);
    }

    /***************************************************************************************************
     * Miscellaneous class methods
     ***************************************************************************************************/
    public void ShowCard()
    {   // Display an instance of a PlayingCard
        // C#: Console.WriteLine() to display to screen

        //  object.method() - C#: Console is an object defined by C3; WriteLine is a method() for Console
        // Console.WriteLine(this.ToString());

        //       class.object.method - System is class defined by Java; out is an object defined by Java
        //                             println() is method defined in the System class (or subclass)
        // Java: System.out.printlin()
        System.out.println(this.toString());
    }

    public void ShowCardWithHash()
    {   // Display an instance of a PlayingCard with HashCode

        //Console.WriteLine(this.ToString() + "\thashCode: " + GetHashCode());
        System.out.println(this.toString() + "\thashCode: " + hashCode());
    }
}  // End of PlayingCard Class